README.md
# oib-tafel

This project is generated with [yo angular generator](https://github.com/yeoman/generator-angular)
version 0.15.1.

## Build & development

Run `grunt` for building and `grunt serve` for preview.

## Testing

Running `grunt test` will run the unit tests with karma.


backend/ChangeLog
Thu 07.01.2016 22:32  Samuel Pulfer <samuel.pulfer@usb.ch>

  * httpd.py, lib/usbauth.py, lib/webctx.py, template/authxmpl.html,
  template/index.html: authorisation example added

Thu 07.01.2016 10:10  wunderlins <simon.wunderlin@usb.ch>

  * bin/clean.sh, bin/setup.sh, bin/start.sh, bin/stop.sh, etc/config.py,
  lib/setenv.sh, lib/web, lib/wsgilog: fixed setup routine

Mon 02.11.2015 08:30  Simon Wunderlin <simon.wunderlin@usb.ch>

  * .gitignore, bin/changelog.py: added changelog script

Mon 02.11.2015 07:50  Simon Wunderlin <simon.wunderlin@usb.ch>

  * bin/restart.sh, bin/start.sh, etc/config.py: fixed bug with paths

Mon 02.11.2015 07:44  Simon Wunderlin <simon.wunderlin@usb.ch>

  * README.md, bin/.gitignore, bin/setup.sh, bin/setuser.py, bin/start.sh,
  bin/status.sh, bin/stop.sh, etc/.gitignore, etc/config.py, etc/user.sql,
  lib/realpath.c, lib/setenv.sh: added setenv and more robust start/stop
  environment

Sat 31.10.2015 23:42  wus <swunderlin@gmail.com>

  * bin/setuser.py, etc/.gitignore, etc/config.py, etc/user.db, static/ext:
  added user mgmnt script

Sat 31.10.2015 23:37  wus <swunderlin@gmail.com>

  * bin/clean.sh, bin/setup.sh, etc/config.py, etc/user.db, httpd.py,
  static/.gitignore, var/.gitignore: tigthened security

Sat 31.10.2015 22:01  wus <swunderlin@gmail.com>

  * bin/clean.sh, bin/genuuid.py, bin/setup.sh, bin/start.sh, bin/stop.sh,
  etc/config.py, httpd.py: moved meta data to config file. added interactive
  setup script. ne cleanup script for testing setup mechanism. genuuid is a
  minimal python script to create a GIUD

Thu 22.10.2015 13:31  Simon Wunderlin <simon.wunderlin@usb.ch>

  * etc/config.py, static/css/login.css, static/js/login.js,
  template/login.html: cosmetics

Wed 21.10.2015 22:11  Samuel Pulfer <samuelpulfer@gmail.com>

  * bin/setup.sh, lib/bootstrap-3.3.5-dist.zip, static/.gitignore,
  static/bootstrap/3.3.5/css/bootstrap-theme.css,
  static/bootstrap/3.3.5/css/bootstrap-theme.css.map,
  static/bootstrap/3.3.5/css/bootstrap-theme.min.css,
  static/bootstrap/3.3.5/css/bootstrap.css,
  static/bootstrap/3.3.5/css/bootstrap.css.map,
  static/bootstrap/3.3.5/css/bootstrap.min.css,
  .../3.3.5/fonts/glyphicons-halflings-regular.eot,
  .../3.3.5/fonts/glyphicons-halflings-regular.svg,
  .../3.3.5/fonts/glyphicons-halflings-regular.ttf,
  .../3.3.5/fonts/glyphicons-halflings-regular.woff,
  .../3.3.5/fonts/glyphicons-halflings-regular.woff2,
  static/bootstrap/3.3.5/js/bootstrap.js,
  static/bootstrap/3.3.5/js/bootstrap.min.js, static/bootstrap/3.3.5/js/npm.js,
  template/bootstrap.html, template/login.html: bootstrap nun als .zip welches
  mit setup.sh entpackt wird.

Tue 20.10.2015 23:57  Samuel Pulfer <samuelpulfer@gmail.com>

  * lib/webctx.py, static/css/login.css, template/login.html: Login Seite
  Bootstrapifiziert.

Tue 20.10.2015 22:54  Samuel Pulfer <samuelpulfer@gmail.com>

  * httpd.py, lib/webctx.py, static/bootstrap/3.3.5/css/bootstrap-theme.css,
  static/bootstrap/3.3.5/css/bootstrap-theme.css.map,
  static/bootstrap/3.3.5/css/bootstrap-theme.min.css,
  static/bootstrap/3.3.5/css/bootstrap.css,
  static/bootstrap/3.3.5/css/bootstrap.css.map,
  static/bootstrap/3.3.5/css/bootstrap.min.css,
  .../3.3.5/fonts/glyphicons-halflings-regular.eot,
  .../3.3.5/fonts/glyphicons-halflings-regular.svg,
  .../3.3.5/fonts/glyphicons-halflings-regular.ttf,
  .../3.3.5/fonts/glyphicons-halflings-regular.woff,
  .../3.3.5/fonts/glyphicons-halflings-regular.woff2,
  static/bootstrap/3.3.5/js/bootstrap.js,
  static/bootstrap/3.3.5/js/bootstrap.min.js, static/bootstrap/3.3.5/js/npm.js,
  static/css/bootstraptheme.css, template/bootstrap.html, template/index.html:
  bootstrap integriert.

Tue 20.10.2015 22:17  Samuel Pulfer <samuelpulfer@gmail.com>

  * static/css/json1.css, static/css/json2.css, static/jquery-2.1.3.js,
  static/jquery-2.1.3.min.js, static/js/.gitignore, static/js/json1.js,
  static/js/json2.js, static/js/login.js, template/json1.html,
  template/json2.html, template/login.html: Grosse Aufraeumaktion. css und js
  aus html entfernt und unter static in entsprechende Ordner geworfen.

Mon 13.07.2015 19:25  wus <swunderlin@gmail.com>

  * etc/config.py, httpd.py, lib/webctx.py, template/error.html,
  test/status.py: fixed auth_check

Fri 10.07.2015 14:30  Simon Wunderlin <simon.wunderlin@usb.ch>

  * httpd.py, lib/webctx.py: fixed session

Fri 10.07.2015 14:24  Simon Wunderlin <simon.wunderlin@usb.ch>

  * httpd.py, lib/webctx.py: moved

Wed 03.06.2015 11:32  Simon Wunderlin <simon.wunderlin@usb.ch>

  * httpd.py, lib/usbauth.py: cleanup

Wed 03.06.2015 10:42  Simon Wunderlin <simon.wunderlin@usb.ch>

  * httpd.py, lib/usbauth.py, template/login.html: added ad authentication

Wed 03.06.2015 09:46  Simon Wunderlin <simon.wunderlin@usb.ch>

  * create_userdb.py, etc/user.db, httpd.py: implemented basic sqlite db aout

Wed 03.06.2015 08:39  Simon Wunderlin <simon.wunderlin@usb.ch>

  * create_userdb.py, etc/user.db, etc/user.sql, httpd.py, lib/auth.py,
  lib/usbauth.py, template/login.html: added simple login form, mechanism, user
  database, needs user authentication adn then the login is done

Tue 02.06.2015 10:28  Simon Wunderlin <simon.wunderlin@usb.ch>

  * httpd.py, lib/auth.py: added auth lib

Tue 10.03.2015 23:00  Samuel Pulfer <samuelpulfer@gmail.com>

  * httpd.py, template/index.html, template/json.html, template/json1.html,
  template/json2.html: Add json example without JQuery. Fix httpd.py (class
  json overrides json include)

Mon 16.02.2015 13:51  Simon Wunderlin <simon.wunderlin@usb.ch>

  * httpd.py, static/py.png: added image example

Fri 13.02.2015 17:20  Simon Wunderlin <simon.wunderlin@usb.ch>

  * template/env.html, var/.gitignore: bug fixes

Fri 13.02.2015 11:41  Simon Wunderlin <simon.wunderlin@usb.ch>

  * httpd.py, static/jquery-2.1.3.js, static/jquery-2.1.3.min.js,
  template/json.html: added jquery, libs

Fri 13.02.2015 10:15  Simon Wunderlin <simon.wunderlin@usb.ch>


test/karma.conf.js
// Karma configuration
// http://karma-runner.github.io/0.12/config/configuration-file.html
// Generated on 2016-02-18 using
// generator-karma 1.0.1

module.exports = function(config) {
  'use strict';

  config.set({
    // enable / disable watching file and executing tests whenever any file changes
    autoWatch: true,

    // base path, that will be used to resolve files and exclude
    basePath: '../',

    // testing framework to use (jasmine/mocha/qunit/...)
    // as well as any additional frameworks (requirejs/chai/sinon/...)
    frameworks: [
      "jasmine"
    ],

    // list of files / patterns to load in the browser
    files: [
      // bower:js
      'bower_components/jquery/dist/jquery.js',
      'bower_components/angular/angular.js',
      'bower_components/bootstrap/dist/js/bootstrap.js',
      'bower_components/angular-animate/angular-animate.js',
      'bower_components/angular-route/angular-route.js',
      'bower_components/angular-mocks/angular-mocks.js',
      // endbower
      "app/scripts/**/*.js",
      "test/mock/**/*.js",
      "test/spec/**/*.js"
    ],

    // list of files / patterns to exclude
    exclude: [
    ],

    // web server port
    port: 8080,

    // Start these browsers, currently available:
    // - Chrome
    // - ChromeCanary
    // - Firefox
    // - Opera
    // - Safari (only Mac)
    // - PhantomJS
    // - IE (only Windows)
    browsers: [
      "PhantomJS"
    ],

    // Which plugins to enable
    plugins: [
      "karma-phantomjs-launcher",
      "karma-jasmine"
    ],

    // Continuous Integration mode
    // if true, it capture browsers, run tests and exit
    singleRun: false,

    colors: true,

    // level of logging
    // possible values: LOG_DISABLE || LOG_ERROR || LOG_WARN || LOG_INFO || LOG_DEBUG
    logLevel: config.LOG_INFO,

    // Uncomment the following lines if you are using grunt's server to run the tests
    // proxies: {
    //   '/': 'http://localhost:9000/'
    // },
    // URL root prevent conflicts with the site root
    // urlRoot: '_karma_'
  });
};


test/spec/controllers/main.js
'use strict';

describe('Controller: MainCtrl', function () {

  // load the controller's module
  beforeEach(module('oibTafelApp'));

  var MainCtrl,
    scope;

  // Initialize the controller and a mock scope
  beforeEach(inject(function ($controller, $rootScope) {
    scope = $rootScope.$new();
    MainCtrl = $controller('MainCtrl', {
      $scope: scope
      // place here mocked dependencies
    });
  }));

  it('should attach a list of awesomeThings to the scope', function () {
    expect(MainCtrl.awesomeThings.length).toBe(3);
  });
});


backend/httpd.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
Start application server on port 8080

The first command line argument will set the port to be bound. Remeber, you 
need root privvileges to bind ports below 1024.

"""

# setup paths
import os, sys
sys.path.append(os.path.join(os.path.dirname(__file__), 'etc'))
sys.path.append(os.path.join(os.path.dirname(__file__), 'lib'))
sys.path.append(os.path.join(os.path.dirname(__file__), 'lib', 'web'))
sys.path.append(os.path.join(os.path.dirname(__file__), 'lib', 'wsgilog'))

# import modules
import web, config, json
import datetime
import time
import tempfile
import sys, logging
from wsgilog import WsgiLog
from cgi import escape
import usbauth
import hashlib
import sqlite3
import webctx

meta = {
	"name": config.meta_name,
	"version": config.meta_version
}

## global variables ############################################################

# url to class mapping
urls = (
  '/mitarbeiter.json', 'webctx.mitarbeiter',
  '/schichtpersonal.json', 'webctx.schichtpersonal',
  '/patienten.json', 'webctx.patienten',
  #'/env', 'webctx.env',
  #'/json1', 'webctx.json1',
  #'/json2', 'webctx.json2',
  #'/image', 'webctx.image',
  #'/login', 'webctx.login',
  #'/bootstrap', 'webctx.bootstrap',
  #'/authorisation', 'webctx.authorisationxmpl'
)

# default session values
session_default = {
	"uid": -1,
	"user": None,
	"email": None
}

## webpy extensions ############################################################

class service(web.application):
	""" allow to pass a custom port/ip into the application """
	def run(self, port=8080, ip='0.0.0.0', *middleware):
		func = self.wsgifunc(*middleware)
		return web.httpserver.runsimple(func, (ip, port))

class Log(WsgiLog):
	""" extend logger, logging to file in var/ """
	def __init__(self, application):
		WsgiLog.__init__(
			self,
			application,
			logformat = '%(message)s',
			tofile = True,
			toprint = True,
			file = config.app_logfile,
			#when = "D",
			#interval = 1,
			#backups = "1000"
		)

class hooks(object):
	@staticmethod
	def load():
		web.debug("Loadhook")
		#web.debug(webctx.session.uid)
		#return "BEGIN"
	
	@staticmethod
	def unload():
		web.debug("Unloadhook")
		#return "ENDE"


# redirect webserver logs to file
#weblog = open(config.web_logfile, "ab")
#sys.stderr = weblog
#sys.stdout = weblog

def init_session(app):
	#global app
	web.debug(web.config.get('_session'))
	if web.config.get('_session') is None:
		web.debug("Setting up new session ...")
		web.config.session_parameters['cookie_name'] = meta["name"]
		web.config.session_parameters['timeout'] = config.session_timeout,
		web.config.session_parameters['secret_key'] = config.session_salt
		web.config.session_parameters['cookie_domain'] = config.session_cookie_domain
		web.config.session_parameters['ignore_expiry'] = config.session_ignore_expiry
		web.config.session_parameters['ignore_change_ip'] = config.session_ignore_change_ip
		web.config.session_parameters['expired_message'] = config.session_expired_message

		temp = tempfile.mkdtemp(dir=config.session_dir, prefix=config.session_dir_prefix)
		webctx.session = web.session.Session(
			app, 
			web.session.DiskStore(temp), 
			initializer = session_default
		)
		#for k in session_default.keys():
		#	webctx.session[k] = session_default[k]
		#webctx.session.uid = -1
		#web.debug(webctx.session.keys())
	else:
		web.debug("Reusing session ...")
		webctx.session = web.config._session
		"""
		try:
			webctx.session["uid"]
		except:
			webctx.session = session_default
		"""
	
	#web.debug("session.uid: %s" % webctx.session.uid) 
	#return webctx.session

## main function ###############################################################

"""
try:
	webctx.session
except NameError:
	web.debug("Resetting session ...")
	webctx.session = None
"""

app = None
if __name__ == "__main__":
	web.config.debug = False

	app = service(urls, globals())
	# session setup, make sure to call it only once if in debug mode
	init_session(app)
	
	app.add_processor(web.loadhook(hooks.load))
	app.add_processor(web.unloadhook(hooks.unload))

	#webctx.session["pid"] += 1
	#print "starting ..."
	#app.add_processor(web.loadhook(loadhook))
	#app.add_processor(web.unloadhook(unloadhook))
	#app.run(config.port, "0.0.0.0")
	
	#webctx.session = get_session(app)

	app.run(config.port, "0.0.0.0", Log)




app/scripts/controllers/main.js
'use strict';

/*
angular.module('oibTafelApp')
	.factory("oibTafelAppFactory", ['$http', function($http) {
	return {
		list: function(callback){
			$http.get('data/patienten.json').success(callback);
		}
	};
}]);
*/

angular.module('oibTafelApp')
	.factory("oibTafelAppFactoryZimmer", ['$http', function($http) {
	return {
		list: function(callback){
			$http.get('data/betten.json').success(callback);
		}
	};
}]);

angular.module('oibTafelApp')
	.factory("oibTafelAppFactoryPatienten", ['$http', function($http) {
	return {
		list: function(callback){
			$http.get('data/patienten.json').success(callback);
		}
	};
}]);

/**
 * OIB Bett object
 *
 * contains room nuber, patient and nurse data
 */
function Bett(zimmer, bett) {
	this.id      = zimmer + "B" + bett;
	this.zimmer  = zimmer;
	this.bett    = bett;
	this.patient = null;
	this.nurse   = null;
	this.class   = "item disabled";
	this.age     = null;
	return this;
}

function Zimmer(nr, trackt) {
	this.nr = nr;
	this.trackt = trackt;
	this.betten = [];
}

// http://stackoverflow.com/questions/10008050/get-age-from-birthdate
function calculate_age(birth_month,birth_day,birth_year) {
	var today_date = new Date();
	var today_year = today_date.getFullYear();
	var today_month = today_date.getMonth();
	var today_day = today_date.getDate();
	var age = today_year - birth_year;

	if ( today_month < (birth_month - 1)) {
		age--;
	}
	if (((birth_month - 1) == today_month) && (today_day < birth_day)) {
		age--;
	}
	return age;
}

/**
 * @ngdoc function
 * @name oibTafelApp.controller:MainCtrl
 * @description
 * # MainCtrl
 * Controller of the oibTafelApp
 */
angular.module('oibTafelApp')
  .controller('MainCtrl', function ($scope, $timeout,
                                            oibTafelAppFactoryZimmer,
                                            oibTafelAppFactoryPatienten) {
    this.awesomeThings = [
      'HTML5 Boilerplate',
      'AngularJS',
      'Karma'
    ];
    
    // pull data from:
    // - PEP
    // - mdsi
    // - tafelDB (a table currently stored in mdsi)
    //$scope.patienten = oibTafelAppFactory.getPatienten();
    //console.log($scope.patienten);
		
		// hold all rooms with all patients
		$scope.data = {};
		
		// keep track of asynchronous calls
		$scope.loaded = {
			patienten: false,
			zimmer: false			
		};
		
		// async data
		$scope.zimmer = [];
		$scope.patienten = [];
		
		// fetch data
		oibTafelAppFactoryZimmer.list(function(oibTafelAppFactoryZimmer) {
			$scope.zimmer = oibTafelAppFactoryZimmer.zimmer;
			$scope.loaded.zimmer = true;
			
			var last_zimmer_nr = null;
			for (var z in $scope.zimmer) {
				var zimmer = $scope.zimmer[z];
				
				if (zimmer.nr !== last_zimmer_nr) {
					last_zimmer_nr = zimmer.nr;
					var new_zimmmer = new Zimmer(zimmer.nr, zimmer.trackt);
					$scope.data[zimmer.nr] = new_zimmmer;
				}
				
				for (var b in zimmer.betten) {
					var bett = new Bett(zimmer.nr, zimmer.betten[b]);
					//zimmer.betten[b];
					$scope.data[zimmer.nr].betten[zimmer.betten[b]] = bett;
					//console.log(bett);
				}
				
			}
			//console.log($scope.data);
			
		});
		 
		oibTafelAppFactoryPatienten.list(function(oibTafelAppFactoryPatienten) {
			$scope.patienten = oibTafelAppFactoryPatienten.patienten;
			//console.log(oibTafelAppFactoryPatienten.patienten)
			$scope.loaded.patienten = true;
		});
		
		// glue beds and patients together, the key is patienten[n].PD_Bett
		$timeout(function () {
			if ($scope.loaded.patienten && $scope.loaded.zimmer) {
				//console.log($scope.loaded.patienten);
				
				/*
				for (var e in $scope.patienten) {
					var bett = $scope.patienten[e].PD_Bett;
					var ix = $scope.betten_data.indexOf(bett, 0);
					if (ix !== -1) {
						//console.log(ix);
						$scope.betten_data[ix].patient = $scope.patienten[e]; 
					}
				}
				//console.log($scope.data);
				*/
				
				for (var e in $scope.patienten) {
					if (!$scope.patienten[e]) {
						continue;
					}
					var zimmerbett = $scope.patienten[e].PD_Bett;
					if (!zimmerbett) {
						continue;
					}
					var zb_arr = zimmerbett.split("B");
					//console.log(zb_arr);
					//console.log(zb_arr.length);
					if (zb_arr.length !== 2) {
						continue;
					}
					try {
						$scope.data[zb_arr[0]].betten[zb_arr[1]].patient = $scope.patienten[e];
						$scope.data[zb_arr[0]].betten[zb_arr[1]].class = "info";
						// calculate age
						var d = $scope.patienten[e].PD_GebDat;
						var birthday = new Date(d.substr(0, 4), parseInt(d.substr(4, 2))-1, d.substr(6, 2));
						var age = calculate_age(birthday.getMonth(), birthday.getDate(), birthday.getYear())-1900;
						$scope.data[zb_arr[0]].betten[zb_arr[1]].age = age;
						//console.log($scope.patienten[e].PD_GebDat + " " + birthday + " " + age);
						
					} catch (e) {
						//console.log(zb_arr);
						continue;
					}
					//console.log(zb_arr);
					
				}				
				
				$scope.loaded.all = 1;
				console.log($scope.data);
				return true;
			}
		}, 100, this); 

  });


Gruntfile.js
// Generated on 2016-02-18 using generator-angular 0.15.1
'use strict';

// # Globbing
// for performance reasons we're only matching one level down:
// 'test/spec/{,*/}*.js'
// use this if you want to recursively match all subfolders:
// 'test/spec/**/*.js'

module.exports = function (grunt) {

  // Time how long tasks take. Can help when optimizing build times
  require('time-grunt')(grunt);

  // Automatically load required Grunt tasks
  require('jit-grunt')(grunt, {
    useminPrepare: 'grunt-usemin',
    ngtemplates: 'grunt-angular-templates',
    cdnify: 'grunt-google-cdn'
  });

  // Configurable paths for the application
  var appConfig = {
    app: require('./bower.json').appPath || 'app',
    dist: 'dist'
  };

  // Define the configuration for all the tasks
  grunt.initConfig({

    // Project settings
    yeoman: appConfig,

    // Watches files for changes and runs tasks based on the changed files
    watch: {
      bower: {
        files: ['bower.json'],
        tasks: ['wiredep']
      },
      js: {
        files: ['<%= yeoman.app %>/scripts/{,*/}*.js'],
        tasks: ['newer:jshint:all', 'newer:jscs:all'],
        options: {
          livereload: '<%= connect.options.livereload %>'
        }
      },
      jsTest: {
        files: ['test/spec/{,*/}*.js'],
        tasks: ['newer:jshint:test', 'newer:jscs:test', 'karma']
      },
      styles: {
        files: ['<%= yeoman.app %>/styles/{,*/}*.css'],
        tasks: ['newer:copy:styles', 'postcss']
      },
      gruntfile: {
        files: ['Gruntfile.js']
      },
      livereload: {
        options: {
          livereload: '<%= connect.options.livereload %>'
        },
        files: [
          '<%= yeoman.app %>/{,*/}*.html',
          '.tmp/styles/{,*/}*.css',
          '<%= yeoman.app %>/images/{,*/}*.{png,jpg,jpeg,gif,webp,svg}'
        ]
      }
    },

    // The actual grunt server settings
    connect: {
      options: {
        port: 9000,
        // Change this to '0.0.0.0' to access the server from outside.
        hostname: '0.0.0.0',
        livereload: 35729
      },
      livereload: {
        options: {
          open: true,
          middleware: function (connect) {
            return [
              connect.static('.tmp'),
              connect().use(
                '/bower_components',
                connect.static('./bower_components')
              ),
              connect().use(
                '/app/styles',
                connect.static('./app/styles')
              ),
              connect.static(appConfig.app)
            ];
          }
        }
      },
      test: {
        options: {
          port: 9001,
          middleware: function (connect) {
            return [
              connect.static('.tmp'),
              connect.static('test'),
              connect().use(
                '/bower_components',
                connect.static('./bower_components')
              ),
              connect.static(appConfig.app)
            ];
          }
        }
      },
      dist: {
        options: {
          open: true,
          base: '<%= yeoman.dist %>'
        }
      }
    },

    // Make sure there are no obvious mistakes
    jshint: {
      options: {
        jshintrc: '.jshintrc',
        reporter: require('jshint-stylish')
      },
      all: {
        src: [
          'Gruntfile.js',
          '<%= yeoman.app %>/scripts/{,*/}*.js'
        ]
      },
      test: {
        options: {
          jshintrc: 'test/.jshintrc'
        },
        src: ['test/spec/{,*/}*.js']
      }
    },

    // Make sure code styles are up to par
    jscs: {
      options: {
        config: '.jscsrc',
        verbose: true
      },
      all: {
        src: [
          'Gruntfile.js',
          '<%= yeoman.app %>/scripts/{,*/}*.js'
        ]
      },
      test: {
        src: ['test/spec/{,*/}*.js']
      }
    },

    // Empties folders to start fresh
    clean: {
      dist: {
        files: [{
          dot: true,
          src: [
            '.tmp',
            '<%= yeoman.dist %>/{,*/}*',
            '!<%= yeoman.dist %>/.git{,*/}*'
          ]
        }]
      },
      server: '.tmp'
    },

    // Add vendor prefixed styles
    postcss: {
      options: {
        processors: [
          require('autoprefixer-core')({browsers: ['last 1 version']})
        ]
      },
      server: {
        options: {
          map: true
        },
        files: [{
          expand: true,
          cwd: '.tmp/styles/',
          src: '{,*/}*.css',
          dest: '.tmp/styles/'
        }]
      },
      dist: {
        files: [{
          expand: true,
          cwd: '.tmp/styles/',
          src: '{,*/}*.css',
          dest: '.tmp/styles/'
        }]
      }
    },

    // Automatically inject Bower components into the app
    wiredep: {
      app: {
        src: ['<%= yeoman.app %>/index.html'],
        ignorePath:  /\.\.\//
      },
      test: {
        devDependencies: true,
        src: '<%= karma.unit.configFile %>',
        ignorePath:  /\.\.\//,
        fileTypes:{
          js: {
            block: /(([\s\t]*)\/{2}\s*?bower:\s*?(\S*))(\n|\r|.)*?(\/{2}\s*endbower)/gi,
              detect: {
                js: /'(.*\.js)'/gi
              },
              replace: {
                js: '\'{{filePath}}\','
              }
            }
          }
      }
    }, 

    // Renames files for browser caching purposes
    filerev: {
      dist: {
        src: [
          '<%= yeoman.dist %>/scripts/{,*/}*.js',
          '<%= yeoman.dist %>/styles/{,*/}*.css',
          '<%= yeoman.dist %>/images/{,*/}*.{png,jpg,jpeg,gif,webp,svg}',
          '<%= yeoman.dist %>/styles/fonts/*'
        ]
      }
    },

    // Reads HTML for usemin blocks to enable smart builds that automatically
    // concat, minify and revision files. Creates configurations in memory so
    // additional tasks can operate on them
    useminPrepare: {
      html: '<%= yeoman.app %>/index.html',
      options: {
        dest: '<%= yeoman.dist %>',
        flow: {
          html: {
            steps: {
              js: ['concat', 'uglifyjs'],
              css: ['cssmin']
            },
            post: {}
          }
        }
      }
    },

    // Performs rewrites based on filerev and the useminPrepare configuration
    usemin: {
      html: ['<%= yeoman.dist %>/{,*/}*.html'],
      css: ['<%= yeoman.dist %>/styles/{,*/}*.css'],
      js: ['<%= yeoman.dist %>/scripts/{,*/}*.js'],
      options: {
        assetsDirs: [
          '<%= yeoman.dist %>',
          '<%= yeoman.dist %>/images',
          '<%= yeoman.dist %>/styles'
        ],
        patterns: {
          js: [[/(images\/[^''""]*\.(png|jpg|jpeg|gif|webp|svg))/g, 'Replacing references to images']]
        }
      }
    },

    // The following *-min tasks will produce minified files in the dist folder
    // By default, your `index.html`'s <!-- Usemin block --> will take care of
    // minification. These next options are pre-configured if you do not wish
    // to use the Usemin blocks.
    // cssmin: {
    //   dist: {
    //     files: {
    //       '<%= yeoman.dist %>/styles/main.css': [
    //         '.tmp/styles/{,*/}*.css'
    //       ]
    //     }
    //   }
    // },
    // uglify: {
    //   dist: {
    //     files: {
    //       '<%= yeoman.dist %>/scripts/scripts.js': [
    //         '<%= yeoman.dist %>/scripts/scripts.js'
    //       ]
    //     }
    //   }
    // },
    // concat: {
    //   dist: {}
    // },

    imagemin: {
      dist: {
        files: [{
          expand: true,
          cwd: '<%= yeoman.app %>/images',
          src: '{,*/}*.{png,jpg,jpeg,gif}',
          dest: '<%= yeoman.dist %>/images'
        }]
      }
    },

    svgmin: {
      dist: {
        files: [{
          expand: true,
          cwd: '<%= yeoman.app %>/images',
          src: '{,*/}*.svg',
          dest: '<%= yeoman.dist %>/images'
        }]
      }
    },

    htmlmin: {
      dist: {
        options: {
          collapseWhitespace: true,
          conservativeCollapse: true,
          collapseBooleanAttributes: true,
          removeCommentsFromCDATA: true
        },
        files: [{
          expand: true,
          cwd: '<%= yeoman.dist %>',
          src: ['*.html'],
          dest: '<%= yeoman.dist %>'
        }]
      }
    },

    ngtemplates: {
      dist: {
        options: {
          module: 'oibTafelApp',
          htmlmin: '<%= htmlmin.dist.options %>',
          usemin: 'scripts/scripts.js'
        },
        cwd: '<%= yeoman.app %>',
        src: 'views/{,*/}*.html',
        dest: '.tmp/templateCache.js'
      }
    },

    // ng-annotate tries to make the code safe for minification automatically
    // by using the Angular long form for dependency injection.
    ngAnnotate: {
      dist: {
        files: [{
          expand: true,
          cwd: '.tmp/concat/scripts',
          src: '*.js',
          dest: '.tmp/concat/scripts'
        }]
      }
    },

    // Replace Google CDN references
    cdnify: {
      dist: {
        html: ['<%= yeoman.dist %>/*.html']
      }
    },

    // Copies remaining files to places other tasks can use
    copy: {
      dist: {
        files: [{
          expand: true,
          dot: true,
          cwd: '<%= yeoman.app %>',
          dest: '<%= yeoman.dist %>',
          src: [
            '*.{ico,png,txt}',
            '*.html',
            'images/{,*/}*.{webp}',
            'styles/fonts/{,*/}*.*'
          ]
        }, {
          expand: true,
          cwd: '.tmp/images',
          dest: '<%= yeoman.dist %>/images',
          src: ['generated/*']
        }, {
          expand: true,
          cwd: 'bower_components/bootstrap/dist',
          src: 'fonts/*',
          dest: '<%= yeoman.dist %>'
        }]
      },
      styles: {
        expand: true,
        cwd: '<%= yeoman.app %>/styles',
        dest: '.tmp/styles/',
        src: '{,*/}*.css'
      }
    },

    // Run some tasks in parallel to speed up the build process
    concurrent: {
      server: [
        'copy:styles'
      ],
      test: [
        'copy:styles'
      ],
      dist: [
        'copy:styles',
        'imagemin',
        'svgmin'
      ]
    },

    // Test settings
    karma: {
      unit: {
        configFile: 'test/karma.conf.js',
        singleRun: true
      }
    }
  });


  grunt.registerTask('serve', 'Compile then start a connect web server', function (target) {
    if (target === 'dist') {
      return grunt.task.run(['build', 'connect:dist:keepalive']);
    }

    grunt.task.run([
      'clean:server',
      'wiredep',
      'concurrent:server',
      'postcss:server',
      'connect:livereload',
      'watch'
    ]);
  });

  grunt.registerTask('server', 'DEPRECATED TASK. Use the "serve" task instead', function (target) {
    grunt.log.warn('The `server` task has been deprecated. Use `grunt serve` to start a server.');
    grunt.task.run(['serve:' + target]);
  });

  grunt.registerTask('test', [
    'clean:server',
    'wiredep',
    'concurrent:test',
    'postcss',
    'connect:test',
    'karma'
  ]);

  grunt.registerTask('build', [
    'clean:dist',
    'wiredep',
    'useminPrepare',
    'concurrent:dist',
    'postcss',
    'ngtemplates',
    'concat',
    'ngAnnotate',
    'copy:dist',
    'cdnify',
    'cssmin',
    'uglify',
    'filerev',
    'usemin',
    'htmlmin'
  ]);

  grunt.registerTask('default', [
    'newer:jshint',
    'newer:jscs',
    'test',
    'build'
  ]);
};
